### PR TITLE
Sparse attention and outputs

### PR DESCRIPTION
**Pull request by bpopeters**
This pull request implements two new features that encourage sparsity in sequence-to-sequence models. First, sparsemax can now be used as an alternative to softmax in computing global attention (see #841 for more description of this). Sparse attention can be used by specifying `-global_attention_function sparsemax` at training time.

Second, this PR introduces a new loss function, `SparsemaxLoss`, for training models to generate sparse probability distributions over the target vocabulary. This is mostly unexplored in the literature, but anecdotally seems to work reasonably well on my favorite toy problem, grapheme-to-phoneme conversion. Sparse output probabilities can be set with `-generator_function sparsemax`at training time. Using sparse outputs could lead to increased efficiency in beam search translation because hypotheses with a log probability of `-inf` can be pruned regardless of beam size. These modifications to the beam search code are left as a future exercise.

In order to make it easy to extend `NMTLossCompute` for new loss functions, I did a little bit refactoring. The code is cleaner and clearer now.

### CODE REVIEW

**Comment by srush** [comment]
This is really cool. Going to try to get it merged this weekend.

**Comment by da03** [comment]
Wow this is so cool, gna run on IWSLT to check the performance.

**Comment by vince62s** [comment]
great any indication on speed perf for this task ?

**Comment by da03** [comment]
Sure sparse attention has a similar speed as the vanilla model (vanilla takes 11982s to finish default command, while sparse attention takes 13898s). Sparse loss takes longer (21632s, I think it's due to the sorting of vocab_size terms per target word generation).

**Comment by vince62s** [comment]
so the only benefit would be at inference ?

**Comment by rbommasanimozilla** [comment]
Is there an implementation reason for why we don't see a speedup since shouldn't backprop for sparse attention be sublinear (as given in the original paper) and therefore be faster. Is it the evaluation of Sparsemax that is slower or worse parallelism/other implementation problems, since other work (conversations with Niculae of Niculae and Blondel, NIPS 2016 - https://papers.nips.cc/paper/6926-a-regularized-framework-for-sparse-and-structured-neural-attention) seems to have gotten that sparsemax is also faster in practice in training.

**Comment by bpopeters** [comment]
I would guess that the slowdown has more to do with the O(n log n) forward pass than with the backward pass. Apparently the alternatives that avoid sorting the input are slower in practice.

I'm not sure how feasible it is to do the sublinear backprop in practice because support for sparse tensors in pytorch is kind of limited. Therefore the tensors used here are sparse in the mathematical sense but dense in their implementation.

But I can also ask Vlad in person tomorrow to see if he has any ideas for speeding it up.
